lib/composer/package/loader/project_loader.rb

```diff
@@ -0,0 +1,52 @@
+module Composer
+  module Package
+    module Loader
+      # Loads a package from existing configuration
+      class ProjectLoader
+
+        def initialize(loader = nil)
+          if loader.nil?
+            loader = Composer::Package::Loader::HashLoader.new
+          end
+          @loader = loader
+        end
+
+        # Load a project ref as a package
+        # Param:  Project project The target gitlab project.
+        # Param:  Branch|Tag ref The target gitlab project branch/tag.
+        # Param:  Hash config A hash containing addional package configuration.
+        # Returns: Composer::Package::Package
+        def load(project, ref, config)
+
+          config['version'] = parse_version(ref)
+
+          # export distribution package
+          # this works only on public repositories since composer can't handle gitlab oAuth yet.
+          if project.public?
+            config['dist'] = {
+              'url'  => [project.web_url, 'repository', 'archive.zip?ref=' + ref.name].join('/'),
+              'type' => 'zip',
+              'reference' => ref.target
+            }
+          end
+
+          # export source package
+          config['source'] = {
+            'url' => project.url_to_repo,
```

Comment:
As @Lumbendil mentioned public projects expose both distribution (archive url) and source information (gitlab repo url) so one can use the archive over the gitlab repo from composer which is faster and more efficient.

When a project is private, what is exposed is meta about this project such as name, description, source url, reference, version, etc. in order to let composer know that we have a package available at a specific version.

Accessing the package is done via composer by performing a git clone so in private repos this requires gitlab authentication (ssh, ssl, etc). 
see @X-Coder note below